### PR TITLE
issue: 1041107 Fix for flow_tag regression with VLAN trunks

### PR DESCRIPTION
**Pull request by olgk**
Fast path for Flow_tag should also adjust correctly Ethernet header for
802.11Q VLAN packets.

Signed-off-by: Oleg Kuporosov <olegk@mellanox.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1823/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1270/ for details (Mellanox internal link).

File: src/vma/dev/ring_simple.cpp [hunk 1]
```diff
@@ -777,10 +777,17 @@ bool ring_simple::rx_process_buffer(mem_buf_desc_t* p_rx_wc_buf_desc, void* pv_f
 
 		if (likely((si != NULL) && si->flow_tag_enabled())) { 
 			// will process packets with set flow_tag_id and enabled for the socket
+			if (unlikely(p_eth_h->h_proto == htons(ETH_P_8021Q))) {
+				// Handle VLAN header as next protocol
+				transport_header_len = ETH_VLAN_HDR_LEN;
+			}
```
Comment (ilansmith) [inline src/vma/dev/ring_simple.cpp]: I would remove the initialization of transport_header_len from the top of the function and set it when we know what it's supposed to be. So at the top of the function I'd do:

`size_t transport_header_len;`

And here, I'd add:

```
if (p_eth_h->h_proto == htons(ETH_P_8021Q)) {
        ...
} else {
        transport_header_len = ETH_HDR_LEN;
}
```

You should do exactly the same in the non-flowtag case bellow.

File: src/vma/dev/ring_simple.cpp [hunk 2]
```diff
@@ -777,10 +777,17 @@ bool ring_simple::rx_process_buffer(mem_buf_desc_t* p_rx_wc_buf_desc, void* pv_f
 
 		if (likely((si != NULL) && si->flow_tag_enabled())) { 
 			// will process packets with set flow_tag_id and enabled for the socket
+			if (unlikely(p_eth_h->h_proto == htons(ETH_P_8021Q))) {
```
Comment (ilansmith) [inline src/vma/dev/ring_simple.cpp]: I wouldn't use `unlikely` here. If there's a vlan defined - then all the packets will have the vlan interface and the condition becomes `likely`.
In principal, `likely`/`unlikely` should be considered at the code flow level (like when looping - it's `unlikely` that the current iteration is the last one, etc...) and not for user level decisions (such as whether the user is "likely" or not to set a vlan).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1920/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1357/ for details (Mellanox internal link).

**Comment by ilansmith** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1363/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1926/ for details (Mellanox internal link).

**Comment by ilansmith** [comment]
Merged #398.

**Pull request closed without merging**